test/model/internal_link_test.dart

```diff
@@ -0,0 +1,387 @@
+    });
+
+    group('on ambiguous new- or old-style: new wins', () {
+      final streams = [
+        eg.stream(streamId: 1,   name: '311'),
+        eg.stream(streamId: 2,   name: '311-'),
+        eg.stream(streamId: 3,   name: '311-help'),
+        eg.stream(streamId: 311, name: 'collider'),
+      ];
+      const testCases = [
+        ('#narrow/stream/311/',      StreamNarrow(311)),
+        ('#narrow/stream/311-/',     StreamNarrow(311)),
+        ('#narrow/stream/311-help/', StreamNarrow(311)),
+      ];
+      testExpectedNarrows(testCases, streams: streams);
+    });
+
+    group('on old stream link', () {
+      final streams = [
+        eg.stream(streamId: 1, name: 'check'),
+        eg.stream(streamId: 2, name: 'bot testing'),
+        eg.stream(streamId: 3, name: 'check.API'),
+        eg.stream(streamId: 4, name: 'stream'),
+        eg.stream(streamId: 5, name: 'topic'),
+      ];
+      const testCases = [
+        ('#narrow/stream/check/',         StreamNarrow(1)),
+        ('#narrow/stream/bot.20testing/', StreamNarrow(2)),
+        ('#narrow/stream/check.2EAPI/',   StreamNarrow(3)),
+        ('#narrow/stream/stream/',        StreamNarrow(4)),
+        ('#narrow/stream/topic/',         StreamNarrow(5)),
+
+        ('#narrow/stream/check.API/',     null),
+      ];
+      testExpectedNarrows(testCases, streams: streams);
+    });
+  });
+
+  group('parseInternalLink', () {
+    group('topic link parsing', () {
+      final stream = eg.stream(name: "general");
+
+      group('basic', () {
+        String mkUrlString(operand) {
+          return '#narrow/stream/${stream.streamId}-${stream.name}/topic/$operand';
+        }
+        final testCases = [
+          (mkUrlString('(no.20topic)'), TopicNarrow(stream.streamId, '(no topic)')),
+          (mkUrlString('lunch'),        TopicNarrow(stream.streamId, 'lunch')),
+        ];
+        testExpectedNarrows(testCases, streams: [stream]);
+      });
+
+      group('on old topic link, with dot-encoding', () {
+        String mkUrlString(operand) {
+          return '#narrow/stream/${stream.name}/topic/$operand';
+        }
+        final testCases = [
+          (mkUrlString('(no.20topic)'), TopicNarrow(stream.streamId, '(no topic)')),
+          (mkUrlString('google.2Ecom'), TopicNarrow(stream.streamId, 'google.com')),
+          (mkUrlString('google.com'),   null),
+          (mkUrlString('topic.20name'), TopicNarrow(stream.streamId, 'topic name')),
+          (mkUrlString('stream'),       TopicNarrow(stream.streamId, 'stream')),
+          (mkUrlString('topic'),        TopicNarrow(stream.streamId, 'topic')),
+        ];
+        testExpectedNarrows(testCases, streams: [stream]);
+      });
+    });
+
+    group('DM link parsing', () {
+      testExpectedDmNarrow(String testCase) {
+        final expectedNarrow = DmNarrow.withUsers([1, 2],
+          selfUserId: eg.selfUser.userId);
+        testExpectedNarrows([(testCase, expectedNarrow)], users: [
```

Comment:
Hmm, this looks wrong — it's a "test" function inside another "test" function.

… Oh, aha. This is a function *definition*, not a function call, of `testExpectedDmNarrow`. So that nesting is fine, then.

Let's give this an explicit return type (of `void`). That's good for function definitions in general because it makes the expectations explicit; and then I guess another benefit I hadn't previously appreciated is that it makes them more visually distinct from function calls.

Reply:
Applied to the same to `testExpectedStreamNarrow` above this